dali/pipeline/operator/direct_operator.h

```diff
@@ -0,0 +1,241 @@
+std::vector<std::shared_ptr<TensorList<CPUBackend>>> DirectOperator<CPUBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<CPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+  return Run<CPUBackend, CPUBackend>(inputs, kwargs, shared_thread_pool.get());
+}
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<GPUBackend>>> DirectOperator<GPUBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<GPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+  return Run<GPUBackend, GPUBackend>(inputs, kwargs, shared_cuda_stream);
+}
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<GPUBackend>>> DirectOperator<MixedBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<CPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+  return Run<CPUBackend, GPUBackend>(inputs, kwargs, shared_cuda_stream);
+}
+
+template <typename Backend>
+template <typename InBackend, typename OutBackend, typename WSInputType, typename WSOutputType>
+std::vector<std::shared_ptr<TensorList<OutBackend>>> DirectOperator<Backend>::RunImpl(
+    const std::vector<std::shared_ptr<TensorList<InBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+  // Convert and add inputs to the workspace.
+  for (size_t in_idx = 0; in_idx < inputs.size(); ++in_idx) {
+    auto tensor_in = std::make_shared<WSInputType>();
+    tensor_in->ShareData(*inputs[in_idx]);
+
+    // Set default layout for input if not specified.
+    if (tensor_in->GetLayout().empty()) {
+      auto default_layout = op_spec.GetSchema().GetInputLayout(
+          in_idx, tensor_in->shape().sample_dim(), tensor_in->GetLayout());
+      if (!default_layout.empty()) {
+        tensor_in->SetLayout(default_layout);
+      }
+    }
+
+    ws.AddInput(tensor_in);
+  }
+
+  for (auto &arg : kwargs) {
+    ws.AddArgumentInput(arg.first, arg.second);
+  }
+
+  std::vector<OutputDesc> output_desc{};
+  std::vector<std::shared_ptr<TensorList<OutBackend>>> outputs{};
+
+  outputs.reserve(num_outputs);
+
+  for (size_t i = 0; i < num_outputs; ++i) {
+    ws.AddOutput(std::make_shared<WSOutputType>(batch_size));
+  }
+
+  ws.SetBatchSizes(batch_size);
+
+  // Setup outputs.
+  if (op->Setup(output_desc, ws) && op->CanInferOutputs()) {
+    for (size_t i = 0; i < num_outputs; ++i) {
+      ws.template Output<OutBackend>(i).Resize(output_desc[i].shape, output_desc[i].type);
+    }
+  }
+
+  op->Run(ws);
+
+  for (size_t i = 0; i < num_outputs; ++i) {
+    outputs.push_back(AsTensorList<OutBackend>(ws.template OutputPtr<OutBackend>(i)));
+  }
+
+  return outputs;
+}
+
+template <typename Backend>
+std::unique_ptr<ThreadPool> DirectOperator<Backend>::shared_thread_pool =
+    std::make_unique<ThreadPool>(1, 0, false);
```

Comment:
Can't we just have an nullptr here?

Reply:
The idea was that we have some useable thread pool by default (used by all eager operators) and only if user wants to change it they can with `Set[Create]ThreadPool`. By making it `nullptr` here we would force users to explicitly create thread pool before using eager operators which could be inconvenient but at the same time would force them to understand how to set it and would avoid questions about how it is not multithreaded.
I guess we should discuss what should be the default behavior here and with the CUDA stream.

Reply:
Understood.